### PR TITLE
New tests for blocked_actions_protection find that it doesn't activate when expected

### PR DESCRIPTION
**Pull request by leoccyao**
This pr started with me trying to write tests for the recently added blocked_actions_sleep functionality; however, upon finding the tests failing, I found that the api send_request() method was not returning "feedback_required" upon hitting a block (see https://github.com/instagrambot/instabot/issues/1087#issuecomment-542549855). Therefore, a second commit was added to reinstate the return value of "feedback_required" (which was present in an earlier version, but seemed to have been removed somewhere).

### CODE REVIEW

**Comment by madshansens** [comment]
sorry to say but the sleep patterns has been tested for months and does only have very little impact.

File: instabot/api/api.py [hunk 1]
```diff
@@ -454,6 +454,7 @@ def send_request(
                         "ATTENTION!: `feedback_required`"
                         + str(response_data.get("feedback_message"))
                     )
+                    return "feedback_required"
```
Comment (ohld) [inline instabot/api/api.py:525]: Why have you added it back? This is bug and should be removed.

**Comment by leoccyao** [comment]
The code coverage report seems to confirm that, with this PR, the blocked actions protection (and sleep delays written before) seems to be hit instead of missed as it was previously.

File: instabot/api/api.py (hunk 1 above)
Comment (ohld) [reply instabot/api/api.py:525]: You shouldn’t include back this line!

**Review by ohld — CHANGES_REQUESTED** [review]
Please read my comments. Mind this PR: https://github.com/instagrambot/instabot/pull/1083

**Comment by leoccyao** [comment]
I understand that adding this return value of feedback_required breaks the interactive 2FA. However, the blocked actions protection checks for a return value of "feedback_required" to protect against blocks. Without returning "feedback_required", the blocked_actions_protection does not seem to work (always missed in code coverage)

Perhaps place the check for interactive 2FA ahead of the general check for "feedback_required"?

**Comment by ohld** [comment]
hmm, this history around "feedback_required" tells be that current logic is shitcode. May I ask you to refactor it and also cover it with tests? That would be really awesome.

**Comment by reliefs** [comment]
feedback_required is used for block protection, i think many packages relying on this "feedback_required" for their bots, as a protection. So to delete this would break many packages include mine if i updated.

Too many feedback_required in a row = ban/action blocked

File: instabot/api/api.py [hunk 2]
```diff
@@ -449,11 +449,15 @@ def send_request(
...
+                if "feedback_required" in str(
+                        response_data.get("message")):
                     self.logger.error(
                         "ATTENTION!: `feedback_required`"
-                        + str(response_data.get("feedback_message"))
+                        + str(response_data.get(
+                            "feedback_message").encode('utf-8'))
                     )
+                    if not response_data.get("two_factor_required"):
+                        return "feedback_required"
```
Comment (ohld) [inline instabot/api/api.py]: hmmm, the same "feedback_required" here. The same problem as we had this similar return value. Please prove that this return value is important (please take into account the pull request mentioned above) or just delete it.

**Comment by leoccyao** [comment]
Sure, though I would appreciate some ideas on how to refactor it. For example, should the interactive two-factor authentication be refactored into the login() function? Also, since I don't currently use the 2FA feature, can I request someone else to help write tests for the 2-factor authentication? Thanks.

**Comment by ohld** [comment]
Guys, please help with your advise on how to refactor login() method to be more robust and scalable. Thanks

**Comment by madshansens** [comment]
i have seen users ask for 2fa, it's a little niche, but we should have it
to work. But it should not cost breaking already used code.

Remember this package is used by many devs and by breaking this would break
many repos.. highly not suggested to break return feedback_required

Because too many feedback_required = ig blocks you. So somehow this can be
used to avoid blocks.

**Comment by ohld** [comment]
The "feedback_required" return is a bug, check #1083

**Comment by leoccyao** [comment]
Is it possible to merge in my other changes (tests written for blocked actions protection and utf-8 support in response checking), while also keeping the question of how to refactor feedback_required open? 

I understand that this is currently a failing build, but due to the blocked actions protection not kicking in, it has been for a while - the tests just call attention to that. If you look at the test results, blocked actions protection is never activated, even in the cases where it should be.

For example:
image

In this case, the "Your follow action is blocked" should show up, but it doesn't.

**Comment by leoccyao** [comment]
Updated title of pull request from 
"Wrote tests for blocked_actions_sleep, fixed protection not working" 
to
"New tests for blocked_actions_protection find that it doesn't activate when expected"
to better reflect the changes in this PR.

**Comment by ohld** [comment]
I definitely can't merge failing build, sorry.

**Comment by leoccyao** [comment]
The reason why feedback_required was problematic is because the send_request method would exit before getting to the check for 2FA. However, it was also needed, because without it, there is no way to activate blocked actions protection.

This commit adds a second check for 2FA in the login method, after feedback_required is returned. Therefore, if 2FA is needed, even if the send_request method exits with "feedback_required", the interactive 2FA will still activate and the login will still be successful. 

It would be nice to have some tests for 2FA, though.

**Comment by ohld** [comment]
Ok. You've done a great job here! Thank you really much for your efforts and code. I can't wait for your new pull requests.

**Review by ohld — APPROVED** [review]

**Comment by patrickrbrandao** [comment]
I'm still getting `feedback_required`This action was blocked...
Do I need to run some of these new tests?

**Pull request merged**